frame/support/src/storage/types/nmap.rs

```diff
@@ -0,0 +1,813 @@
+			assert_eq!(A::get((2, ())), Some(100));
+
+
+			A::insert((2, ()), 10);
+			assert_eq!(A::take((2, ())), Some(10));
+			assert_eq!(A::contains_key((2, ())), false);
+			assert_eq!(AValueQueryWithAnOnEmpty::take((2, ())), 98);
+			assert_eq!(A::contains_key((2, ())), false);
+			assert_eq!(A::try_get((2, ())), Err(()));
+
+			B::insert((2, ()), 10);
+			assert_eq!(A::migrate_keys::<KeyHasher<Blake2_256>, _>((2, ())), Some(10));
+			assert_eq!(A::contains_key((2, ())), true);
+			assert_eq!(A::get((2, ())), Some(10));
+
+			A::insert((3, ()), 10);
+			A::insert((4, ()), 10);
+			A::remove_all();
+			assert_eq!(A::contains_key((3, ())), false);
+			assert_eq!(A::contains_key((4, ())), false);
+
+			A::insert((3, ()), 10);
+			A::insert((4, ()), 10);
+			assert_eq!(A::iter_values().collect::<Vec<_>>(), vec![10, 10]);
+
+			C::insert((3, ()), 10);
+			C::insert((4, ()), 10);
+			A::translate_values::<u8,_>(|v| Some((v * 2).into()));
+			assert_eq!(A::iter().collect::<Vec<_>>(), vec![((4, ()), 20), ((3, ()), 20)]);
+
+			A::insert((3, ()), 10);
+			A::insert((4, ()), 10);
+			assert_eq!(A::iter().collect::<Vec<_>>(), vec![((4, ()), 10), ((3, ()), 10)]);
+			assert_eq!(A::drain().collect::<Vec<_>>(), vec![((4, ()), 10), ((3, ()), 10)]);
+			assert_eq!(A::iter().collect::<Vec<_>>(), vec![]);
+
+			C::insert((3, ()), 10);
+			C::insert((4, ()), 10);
+			A::translate::<u8,_>(|(k1, ()), v| Some((k1 as u16 * v as u16).into()));
+			assert_eq!(A::iter().collect::<Vec<_>>(), vec![((4, ()), 40), ((3, ()), 30)]);
+
+			assert_eq!(A::MODIFIER, StorageEntryModifier::Optional);
+			assert_eq!(AValueQueryWithAnOnEmpty::MODIFIER, StorageEntryModifier::Default);
+			assert_eq!(A::NAME, "foo");
+			assert_eq!(AValueQueryWithAnOnEmpty::DEFAULT.0.default_byte(), 98u32.encode());
+			assert_eq!(A::DEFAULT.0.default_byte(), Option::<u32>::None.encode());
+
+			WithLen::remove_all();
+			assert_eq!(WithLen::decode_len((3, ())), None);
+			WithLen::append((0, ()), 10);
+			assert_eq!(WithLen::decode_len((0, ())), Some(1));
+		});
+	}
+
+	#[test]
+	fn test_2_keys() {
+		type A = StorageNMap<Prefix, Key<Blake2_128Concat, u16, Key<Twox64Concat, u8>>, u32, OptionQuery>;
+		type AValueQueryWithAnOnEmpty = StorageNMap<
+			Prefix, Key<Blake2_128Concat, u16, Key<Twox64Concat, u8>>, u32, ValueQuery, ADefault
+		>;
+		type B = StorageNMap<Prefix, Key<Blake2_256, u16, Key<Twox128, u8>>, u32, ValueQuery>;
+		type C = StorageNMap<Prefix, Key<Blake2_128Concat, u16, Key<Twox64Concat, u8>>, u8, ValueQuery>;
+		type WithLen = StorageNMap<Prefix, Key<Blake2_128Concat, u16, Key<Twox64Concat, u8>>, Vec<u32>>;
+
+		TestExternalities::default().execute_with(|| {
+			let mut k: Vec<u8> = vec![];
+			k.extend(&twox_128(b"test"));
+			k.extend(&twox_128(b"foo"));
+			k.extend(&3u16.blake2_128_concat());
+			k.extend(&30u8.twox_64_concat());
+			assert_eq!(A::hashed_key_for((3, (30, ()))).to_vec(), k);
+
+			assert_eq!(A::contains_key((3, (30, ()))), false);
```

Comment:
I really don't like the extra `()` that we have to tack on in order for the type system to work... if Rust supports specialization, I could have added another `impl<H1, K1, H2, K2, KG> KeyGenerator for Key<H1, K1, Key<H2, K2, ()>>` that halts the recursion at the second-to-last key, eliminating the extra `()`.

This however still doesn't fix the 1 key case, which still requires the extra `()`, but it would be so much nicer for maps with more than 1 key.

Reply:
It seems there is some possibilities:
I improved the EncodeLike a bit in parity-scale-codec:
```diff
diff --git a/src/codec.rs b/src/codec.rs
index 54fe4c9..6707ea4 100644
--- a/src/codec.rs
+++ b/src/codec.rs
@@ -1089,6 +1089,14 @@ macro_rules! tuple_impl {
        }
 }
 
+impl<F: EncodeLike<FE>, FE: Encode> crate::EncodeLike<(FE, ())> for (F,) {}
+impl<F: EncodeLike<FE>, FE: Encode, G: EncodeLike<GE>, GE: Encode> crate::EncodeLike<(FE, GE, ())> for (F, G) {}
+impl<
+       F: EncodeLike<FE>, FE: Encode,
+       G: EncodeLike<GE>, GE: Encode,
+       H: EncodeLike<HE>, HE: Encode,
+> crate::EncodeLike<(FE, GE, HE, ())> for (F, G, H) {}
+
 #[allow(non_snake_case)]
 mod inner_tuple_impl {
        use super::*;
```

and then I was able to get rid of the extra `()`. Though code is still not very beautiful as it expect a tuple. So I had to write `(1,)` instead of `1`
```diff
diff --git a/frame/support/src/storage/types/key.rs b/frame/support/src/storage/types/key.rs
index 878b5ad872..8a3fa87adb 100755
--- a/frame/support/src/storage/types/key.rs
+++ b/frame/support/src/storage/types/key.rs
@@ -64,8 +64,8 @@ impl<H: StorageHasher, KHG: KeyHasherGenerator> KeyHasherGenerator for KeyHasher
 
 /// A trait that contains the current key as an associated type and generates the next key.
 pub trait KeyGenerator {
-       type CurrentKey: EncodeLike<Self::CurrentKey>;
-       type NextKey;
+       type CurrentKey: codec::Encode + EncodeLike<Self::CurrentKey>;
+       type NextKey: codec::Encode;
 
        fn final_key<KArg: EncodeLike<Self::CurrentKey>>(k: &(KArg, Self::NextKey)) -> Vec<u8>;
        fn final_hash<KArg, KHG>(k: &(KArg, Self::NextKey)) -> Vec<u8>
diff --git a/frame/support/src/storage/types/nmap.rs b/frame/support/src/storage/types/nmap.rs
index 4c37e7ff0e..5539e9f98e 100755
--- a/frame/support/src/storage/types/nmap.rs
+++ b/frame/support/src/storage/types/nmap.rs
@@ -140,12 +140,11 @@ where
        }
 
        /// Store a value to be associated with the given keys from the map.
-       pub fn insert<KArg, VArg>(key: (KArg, Key::NextKey), val: VArg)
+       pub fn insert<KArg, VArg>(key: KArg, val: VArg)
        where
-               KArg: EncodeLike<Key::CurrentKey>,
-               VArg: EncodeLike<Value>,
+               KArg: EncodeLike<(Key::CurrentKey, Key::NextKey)>,
        {
-               <Self as crate::storage::StorageNMap<Key, Value>>::insert(key, val)
+               todo!();
        }
 
        /// Remove the value under the given keys.
@@ -385,7 +384,7 @@ mod test {
                        assert_eq!(A::get((3, ())), None);
                        assert_eq!(AValueQueryWithAnOnEmpty::get((3, ())), 98);
 
-                       A::insert((3, ()), 10);
+                       A::insert((3,), 10);
                        assert_eq!(A::contains_key((3, ())), true);
                        assert_eq!(A::get((3, ())), Some(10));
                        assert_eq!(AValueQueryWithAnOnEmpty::get((3, ())), 10);
```

Reply:
maybe instead of having `NextKey` in KeyGenerator you could have `FullKey` instead, and remove the last `()` in the `FullKey`. I'm not sure yet, I need to dive a bit more, but would be good to get rid of the extra `()`. Even if we can't get rid for storage0map or storage1map it is good to have IMO